src/runtime/local/kernels/FPGAOPENCL/gemm_interface.cpp

```diff
@@ -0,0 +1,701 @@
+#include <iostream>
+#include <math.h>
+#include <sstream>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string>
+#include <sys/time.h>
+#include <time.h>
+
+// Parameters of the systolic array
+#define II   32
+#define JJ   32
+#define KK   32
+#define III  14
+#define JJJ  16
+#define KKK  16
+
+using namespace aocl_utils;
+
+#define TYPE float
+
+#define STR_HELPER(x) #x
+#define STR(x) STR_HELPER(x)
+
+#define DPRINTF(...)     \
+    printf(__VA_ARGS__); \
+    fflush(stdout);
+
+#define NUM_QUEUES_TO_CREATE    6
+#define NUM_KERNELS_TO_CREATE   6
+
+#define CHECK(status)                                       \
+    if (status != CL_SUCCESS) {                             \
+        printf("error %d in line %d.\n", status, __LINE__); \
+        exit(1);                                            \
+    }
+
+#define ACL_ALIGNMENT 64
+void *acl_aligned_malloc(size_t size) {
+    void *result = NULL;
+    posix_memalign(&result, ACL_ALIGNMENT, size);
+    return result;
+}
+
+void cleanup() {}
+
+const char *kernel_name[] = {
+    "kernel_A_loader",
+    "kernel_B_loader",
+    "kernel_unloader_WAIT_FINISH",
+    "kernel_A_feeder",
+    "kernel_B_feeder",
+    "kernel_Out"
+};
+
+double compute_kernel_execution_time(cl_event &event, double &start_d, double &end_d) {
+    cl_ulong start, end;
+
+    clGetEventProfilingInfo(event, CL_PROFILING_COMMAND_END, sizeof(cl_ulong), &end, NULL);
+    clGetEventProfilingInfo(event, CL_PROFILING_COMMAND_START, sizeof(cl_ulong), &start, NULL);
+
+    start_d = (double)1.0e-9 * start;
+    end_d = (double)1.0e-9 * end;
+    //return (double)(end-start);
+    return (double)1.0e-9 * (end - start); // nanoseconds to seconds
+}
+
+
+int sgemm(const float *A, const float *B, float *C, const int OUTERMOST_I, const int OUTERMOST_J, const int OUTERMOST_K) {
+    const int TOTAL_I = III * II * OUTERMOST_I;
+    const int TOTAL_J = JJJ * JJ * OUTERMOST_J;
+    const int TOTAL_K = KKK * KK * OUTERMOST_K;
+    
+    long int num_elem_A = (long int)TOTAL_I*TOTAL_K;
+    long int num_elem_B = (long int)TOTAL_K*TOTAL_J;
+    long int num_elem_C = (long int)TOTAL_I*TOTAL_J;
+
+    float *serialized_A, *serialized_B;
+    if ((serialized_A = (float *)acl_aligned_malloc(num_elem_A * sizeof(float))) == NULL) {
+        perror("Failed malloc of matrix serialized_A");
+    }
+    if ((serialized_B = (float *)acl_aligned_malloc(num_elem_B * sizeof(float))) == NULL) {
+        perror("Failed malloc of matrix serialized_A");
+    }
+
+    // Serialize A
+    long int addr = 0;
+    for (int i = 0; i < TOTAL_I; i++)
+        for (int k = 0; k < TOTAL_K; k++) {
+            serialized_A[addr++] = A[k + i*TOTAL_K];
+        }
+    // Serialize B
+    addr = 0;
+    for (int j = 0; j < TOTAL_J; j++)
+        for (int k = 0; k < TOTAL_K; k++) {
+            serialized_B[addr++] = B[j+k*TOTAL_J];
+        }
+#ifndef NDEBUG
+    DPRINTF("\n===== Host-CPU setting up the OpenCL platform and device ======\n\n");
```

Comment:
All the setup code should be moved to a separate function that is called in the beginning of the Daphne execution, not on every sgemm()

Reply:
any additional details where exactly such function should be called?